### PR TITLE
Add timestamp `stype`

### PR DESCRIPTION
**Pull request by yiweny**

### CODE REVIEW

File: torch_frame/data/mapper.py [hunk 1]
```diff
@@ -211,6 +211,39 @@ def backward(self, tensor: MultiNestedTensor) -> pd.Series:
...
+
+    def forward(
+        self,
+        ser: Series,
+        *,
+        device: Optional[torch.device] = None,
+    ) -> Tensor:
+        ser = pd.to_datetime(ser)
+        tensors = [
+            torch.tensor(ser.dt.year.values, device=device),
```
Comment (rusty1s) [inline torch_frame/data/mapper.py:227]:
```suggestion
            torch.from_numpy(ser.dt.year.values).to(device=device),
```

File: torch_frame/data/mapper.py [hunk 2]
```diff
@@ -211,6 +218,45 @@ def backward(self, tensor: MultiNestedTensor) -> pd.Series:
...
+                NUM_DAYS_PER_MONTH).to(device=device).unsqueeze(1),
+            torch.from_numpy(ser.dt.dayofweek.values /
+                             NUM_DAYS_PER_WEEK).to(device=device).unsqueeze(1),
+            torch.from_numpy(ser.dt.hour.values /
+                             NUM_HOURS_PER_DAY).to(device=device).unsqueeze(1),
+            torch.from_numpy(ser.dt.minute.values / NUM_MINUTES_PER_HOUR).to(
+                device=device).unsqueeze(1),
+            torch.from_numpy(ser.dt.second.values / NUM_SECONDS_PER_MINUTE).to(
+                device=device).unsqueeze(1)
+        ]
```
Comment (weihua916) [inline torch_frame/data/mapper.py:253]: Put the normalization logic into `StypeEncoder`.

File: torch_frame/data/dataset.py [hunk 3]
```diff
@@ -260,6 +281,13 @@ class Dataset(ABC):
             values are tensors such as tokens.
             :obj:`batch_size` specifies the mini-batch size for
             :obj:`text_tokenizer`. (default: :obj:`None`)
+        col_to_time_format (Union[str, Dict[str, str]], optional): A
+            dictionary or a string specifying the format for the timestamp
+            columns. If a string is specified, then the same separator will
+            be used throughout all the multi-categorical columns. If a
+            dictionary is given, we use a separator specified for each
+            column. If not specified, pandas's internal to_datetime function
+            will be used to auto parse time columns. (default: None)
```
Comment (weihua916) [inline torch_frame/data/dataset.py:290]: It'd be great if you can provide a link on how to write `time_format`. Also, the doc-string seems to contain multi-categorical information. Can you re-write?

File: torch_frame/data/dataset.py [hunk 4]
```diff
@@ -179,6 +195,11 @@ def _get_mapper(self, col: str) -> TensorMapper:
             index, _ = self.col_stats[col][StatType.MULTI_COUNT]
             return MultiCategoricalTensorMapper(index,
                                                 sep=self.col_to_sep[col])
+        elif stype == torch_frame.timestamp:
+            year_range = self.col_stats[col][StatType.YEAR_RANGE]
```
Comment (weihua916) [inline torch_frame/data/dataset.py:199]: This `year_range` should be passed to `StypeEncoder`.

File: test/data/test_mapper.py [hunk 5]
```diff
@@ -41,6 +47,32 @@ def test_categorical_tensor_mapper():
...
+                                   time_format=format)[StatType.YEAR_RANGE]
+
+    mapper = TimestampTensorMapper(format=format, year_range=year_range)
+
+    out = mapper.forward(ser)
+    assert out.shape == (num_rows, 1, 7)
+    assert torch.isnan(out[0, :, :]).all()
+    assert out[1:, :, :].dtype == torch.float32
+    is_between_0_and_1 = (out[1:, :, :] >= 0) & (out[1:, :, :] < 1)
+    assert is_between_0_and_1.all()
```
Comment (weihua916) [inline test/data/test_mapper.py:73]: I think we can generate some specified timestamp (not random), and test if we can read off the correct timestamp from the PyTorch Tensor.

File: torch_frame/data/stats.py [hunk 6]
```diff
@@ -85,20 +96,27 @@ def compute(self, ser: Series, sep: Optional[str] = None) -> Any:
...
+
 
 _default_values = {
     StatType.MEAN: np.nan,
     StatType.STD: np.nan,
     StatType.QUANTILES: [np.nan, np.nan, np.nan, np.nan, np.nan],
     StatType.COUNT: ([], []),
-    StatType.MULTI_COUNT: ([], [])
+    StatType.MULTI_COUNT: ([], []),
+    StatType.YEAR_RANGE: [1900, 1900],
```
Comment (weihua916) [inline torch_frame/data/stats.py:111]: np.nan

**Review by weihua916 — COMMENTED** [review]
Thank you. This is great. Left some comments.

File: test/data/test_mapper.py [hunk 7]
```diff
@@ -41,6 +45,30 @@ def test_categorical_tensor_mapper():
...
+
+    mapper = TimestampTensorMapper(format=format)
+
+    out = mapper.forward(ser)
+    assert out.shape == (num_rows, 1, 7)
+    assert torch.isnan(out[0, :, :]).all()
+    assert torch.allclose(
+        out[1, :, :],
+        torch.tensor([2020., 3., 9., 0, 17., 20., 4.]).view(1, -1))
+    assert out[1:, :, :].dtype == torch.float32
```
Comment (weihua916) [inline test/data/test_mapper.py:69]: nit:
```
assert out.dtype == torch.float32
```

Comment (weihua916) [inline test/data/test_mapper.py:69]: `out[0]`, `out[1]` are enough.

File: torch_frame/data/dataset.py [hunk 8]
```diff
@@ -149,10 +162,16 @@ def __init__(
...
-        self.col_to_sep = canonicalize_col_to_sep(
+        self.col_to_sep = canonicalize_col_to_pattern(
             col_to_sep,
             self.col_names_dict.get(torch_frame.multicategorical, []))
 
+        self.col_to_time_format = (None if col_to_time_format is None else
+                                   canonicalize_col_to_pattern(
+                                       col_to_time_format,
+                                       self.col_names_dict.get(
+                                           torch_frame.timestamp, [])))
```
Comment (weihua916) [inline torch_frame/data/dataset.py:173]: In this canonicalization, can we transform `None` to 
`{'col1': None, 'col2': None}` as well?

File: torch_frame/data/dataset.py [hunk 9]
```diff
@@ -179,6 +198,11 @@ def _get_mapper(self, col: str) -> TensorMapper:
             index, _ = self.col_stats[col][StatType.MULTI_COUNT]
             return MultiCategoricalTensorMapper(index,
                                                 sep=self.col_to_sep[col])
+        elif stype == torch_frame.timestamp:
+            year_range = self.col_stats[col][StatType.YEAR_RANGE]
+            return TimestampTensorMapper(
+                format=(None if self.col_to_time_format is None else
+                        self.col_to_time_format[col]), year_range=year_range)
```
Comment (weihua916) [inline torch_frame/data/dataset.py]: `self.col_to_time_format[col]` is enough if we canonicalize.

File: torch_frame/data/dataset.py [hunk 10]
```diff
@@ -303,10 +338,16 @@ def __init__(
...
+        self.col_to_sep = canonicalize_col_to_pattern(col_to_sep, [
             col for col, stype in self.col_to_stype.items()
             if stype == torch_frame.multicategorical
         ])
+        self.col_to_time_format = (
+            None if col_to_time_format is None else
+            canonicalize_col_to_pattern(col_to_time_format, [
+                col for col, stype in self.col_to_stype.items()
+                if stype == torch_frame.timestamp
+            ]))
```
Comment (weihua916) [inline torch_frame/data/dataset.py:349]: Same here. Let's canonicalize `None` as well. The output here should always be a dictionary mapping from col name into time_format (which can be `None`).

File: torch_frame/data/dataset.py [hunk 11]
```diff
@@ -428,7 +469,9 @@ def materialize(self, device: Optional[torch.device] = None,
         for col, stype in self.col_to_stype.items():
             ser = self.df[col]
             self._col_stats[col] = compute_col_stats(
-                ser, stype, sep=self.col_to_sep.get(col, None))
+                ser, stype, sep=self.col_to_sep.get(col, None),
+                time_format=None if self.col_to_time_format is None else
+                self.col_to_time_format.get(col, None))
```
Comment (weihua916) [inline torch_frame/data/dataset.py:473]: `self.col_to_time_format.get(col, None)` alone is enough.

File: test/data/test_stats.py [hunk 12]
```diff
@@ -36,6 +40,49 @@ def test_compute_col_stats_all_multi_categorical():
...
+        _random_timestamp(start_date, end_date, format)
+        for _ in range(num_rows)
+    ]
+    arr[0::2] = len(arr[0::2]) * [np.nan]
+    ser = pd.Series(arr)
+    stype = timestamp
+
+    year_range = compute_col_stats(ser, stype,
+                                   time_format=format)[StatType.YEAR_RANGE]
+    assert (year_range[0] == year_range[1] == 1900)
```
Comment (weihua916) [inline test/data/test_stats.py:75]: why is this 1900?

File: torch_frame/data/dataset.py [hunk 13]
```diff
@@ -57,36 +58,38 @@ def _requires_post_materialization(self, *args, **kwargs):
...
     """
-    if isinstance(col_to_sep, str):
-        sep = col_to_sep
-        col_to_sep = {}
+    if col_to_pattern is None or isinstance(col_to_pattern, str):
+        sep = col_to_pattern
+        col_to_pattern = {}
         for col in columns:
-            col_to_sep[col] = sep
+            col_to_pattern[col] = sep
```
Comment (weihua916) [inline torch_frame/data/dataset.py:84]: nit. `sep` --> `pattern`

**Review by weihua916 — APPROVED** [review]
LGTM, thanks! final nit.
